pkg/util/assertion/assertion.go

```diff
@@ -0,0 +1,219 @@
+// budget isn't closed with outstanding allocations. In release builds, we
+// simply release the outstanding bytes and continue closing.
+//
+//	func (m *BytesMonitor) Close(ctx context.Context) {
+//	  m.mu.Lock()
+//	  defer m.mu.Unlock()
+//	  if m.mu.curAllocated != 0 {
+//	    _ = assertion.Failed(ctx, "unexpected %d leftover bytes", m.mu.curAllocated)
+//	    m.releaseBytesLocked(ctx, m.mu.curAllocated) // runs in release builds
+//	  }
+//	  mm.releaseBudget(ctx)
+//	}
+//
+// Example: returning an error on assertion failure. Here, we guard against
+// unknown transaction statuses during EndTxn. In release builds, we return an
+// error to the RPC client.
+//
+//	func EndTxn(
+//	  ctx context.Context, rw storage.ReadWriter, cArgs CommandArgs, resp kvpb.Response,
+//	) (result.Result, error) {
+//	  switch reply.Txn.Status {
+//	  case roachpb.COMMITTED: // ...
+//	  case roachpb.ABORTED:   // ...
+//	  case roachpb.PENDING:   // ...
+//	  case roachpb.STAGING:   // ...
+//	  default:
+//	    err := assertion.Failed(ctx, "invalid txn status %s", reply.Txn.Status)
+//	    return result.Result{}, err // runs in release builds
+//	  }
+//	}
+//
+// Example: expensive assertions. Here, we assert that a range's MVCC stats are
+// accurate after every write, verifying that incremental updates done by writes
+// are correct. This is too expensive to do even in non-release builds, so we
+// gate the assertion on ExpensiveEnabled which is only enabled in special test
+// builds.
+//
+//	func (b *replicaAppBatch) ApplyToStateMachine(ctx context.Context) error {
+//	  // ...
+//	  if assertion.ExpensiveEnabled {
+//	    // runs in certain test builds
+//	    stats, err := storage.ComputeStats(b.r.store.Engine, desc.StartKey, desc.EndKey, now)
+//	    if err != nil {
+//	      return err
+//	    }
+//	    if stats != *b.state.Stats {
+//	      return assertion.Failed(ctx, "incorrect MVCC stats, %v != %v", stats, b.state.Stats)
+//	    }
+//	  }
+//	  // ...
+//	}
+//
+// Example: fatal errors, which should not be assertions. Here, Pebble detects
+// file corruption. This is not an assertion, but rather normal error handling,
+// so we fatal via log.Fatalf() even in release builds.
+//
+//	if err := iter.Close(); err == pebble.ErrCorruption {
+//	  log.Fatalf(ctx, "%v", err)
+//	}
+package assertion
+
+import (
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/build"
+	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/envutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+var (
+	// shouldFatal exits the process with a fatal error on assertion failures. It
+	// defaults to true in non-release builds.
+	shouldFatal = envutil.EnvOrDefaultBool("COCKROACH_FATAL_ASSERTIONS", !build.IsRelease())
+
+	// MaybeSendReport is injected by package logcrash, for Sentry reporting.
+	MaybeSendReport func(ctx context.Context, err error)
```

Comment:
So, in all builds that import `util/log/logcrash` package one way or another (but not in builds that don't), this will report to sentry, unless explicitly overridden? Even from local / test builds?

Or, is the reporting guarded by some other condition (like, this is a release build or something)? By looking at the code briefly, it seems like we're not always reporting.

Reply:
>So, in all builds that import util/log/logcrash package one way or another (but not in builds that don't), this will report to sentry, unless explicitly overridden?

It's enabled in binaries. The function is injected when importing `logcrash`, but the reporting is gated on a default-true setting, and it relies on injecting a settings object that's done by the `cli` package. This is the same logic `log.Fatalf` already uses. See this code:

https://github.com/cockroachdb/cockroach/blob/c7ff5ee0d71a02f02e17ee02d28b5c5c796332d0/pkg/util/log/logcrash/crash_reporting.go#L453-L458